contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,158 @@
+
+    /// @notice tracks block, previously claimed fees,
+    /// kwenta balance of this contract, and total
+    /// staked amount in StakingRewardsV2
+    struct Distribution {
+        uint epochStartBlockNumber;
+        uint epochStartTime;
+        uint previouslyClaimedFees;
+        uint kwentaStartOfEpoch;
+    }
+
+    /// @notice tracks the distribution for each epoch
+    mapping(uint => Distribution) public distributionEpochs;
+
+    /// @notice represents the status of if a person already
+    /// claimed their epoch
+    mapping(address => mapping(uint => bool)) public claimedEpochs;
+
+    /// @notice kwenta interface
+    IKwenta public kwenta;
+
+    /// @notice Counter for new epochs (starts at 0)
+    uint256 public epoch;
+
+    /// @notice running total for claimed fees
+    uint256 public claimedFees;
+
+    /// @notice rewards staking contract
+    StakingRewardsV2 public stakingRewardsV2;
+
+    /// @notice escrow contract which holds (and may stake) reward tokens
+    RewardEscrowV2 public rewardEscrowV2;
+
+    constructor(
+        address _kwenta,
+        address _stakingRewardsV2,
+        address _rewardEscrowV2
+    ) {
+        kwenta = IKwenta(_kwenta);
+        epoch = 0;
+        stakingRewardsV2 = StakingRewardsV2(_stakingRewardsV2);
+        rewardEscrowV2 = RewardEscrowV2(_rewardEscrowV2);
+    }
+
+    /// @notice  creates a new Distribution entry at the current block,
+    /// can only be called once per week
+    function newDistribution() public {
+        ///@dev [epoch - 1] to get the start of last weeks epoch
+        if (epoch > 0) {
+            require(
+                block.timestamp >=
+                    (distributionEpochs[epoch - 1].epochStartTime + 1 weeks),
+                "TokenDistributor: Last week's epoch has not ended yet"
+            );
+        }
+
+        Distribution memory distribution = Distribution(
+            block.number,
+            block.timestamp,
+            claimedFees,
+            kwenta.balanceOf(address(this))
+        );
+        //for a given block it might not be the final value of the block
+        distributionEpochs[epoch] = distribution;
+        //todo: timestamp is put in for block but that is technically the time
+        emit NewEpochCreated(block.timestamp, epoch);
+
+        epoch++;
+    }
+
+    /// @notice this function will fetch StakingRewardsV2 to see what their staked balance
+    /// was at the start of the epoch then calculate proportional fees and transfer to user
+    function claimDistribution(address to, uint epochNumber) public {
+        /// @dev if this is the first claim of a new epoch, call newDistribution to start a new epoch
+        if (
+            block.timestamp >=
+            (distributionEpochs[epoch - 1].epochStartTime + 604800)
+        ) {
+            newDistribution();
+        }
+        //require the other cases
+        //add custom error (saves gas)
+        require(
+            epochNumber < (epoch - 1),
```

Comment:
This math don't check out

For example, if epoch 0 is posted and someone tries to claim for epoch 0:
```
epochNumber: 0
epoch: 1
(epoch - 1): 0

epochNumber < (epoch - 1) would return false
```

Has this claim distribution function been tested?

Reply:
related to the above comment. Yes its been tested but it functions to the above logic. You can't claim for epoch 0 until epoch 1, but while you are in the middle of epoch 1 the uint256 epoch is actually == 2. So to clarify: epoch-1 is the current epoch you are going through and epoch-2 is the soonest epoch you can claim for. 

The numbering could likely be made less confusing, but it's mostly for the logic that needs data for fees recorded at the START of an epoch. Thinking of it now though I think I know how to fix it (or at least make it clearer).